[FIX] microsoft_calendar: Make searching organizer by their email address case-insensitive

mlaitinen opened the pull request:
Description of the issue/feature this PR addresses:

When a calendar event gets synchronized from Outlook calendar to Odoo in which the organizer's email address' capitalization doesn't match between Outlook and Odoo, the event is created in Odoo without an organizer. (i.e. `user_id` is `False`)

Current behavior before PR:

If the organizer's email address is `Firstname.Lastname@example.com` in Outlook and `firstname.lastname@example.com` in Odoo, the event is created in Odoo without an organizer.

Desired behavior after PR is merged:

Odoo is able to find the correct organizer user ID even though the capitalization of the email address wouldn't match.


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/147069).

arj-odoo reviewed (changes requested):
Note: it could have strange result
jacques@exammple.com
jean-jacques@example.com will both match

```py
In [1]: env['res.partner'].search([('email', '=', 'jacques@example.com')])
Out[1]: res.partner(10,)

In [2]: env['res.partner'].search([('email', '=', 'Jacques@example.com')])
Out[2]: res.partner()

In [3]: env['res.partner'].search([('email', '=', 'jean-Jacques@example.com')])
Out[3]: res.partner(11,)

In [4]: env['res.partner'].search([('email', 'ilike', 'jean-Jacques@example.com')])
Out[4]: res.partner(11,)

In [5]: env['res.partner'].search([('email', 'ilike', 'jacques@example.com')])
Out[5]: res.partner(10, 11)
```
Depending of their order in the database, one or the other will match. 
It could have catastrophic effects where random users are added to events, synchronization push these changes and nobody understand what is going on.
Moreover your change does not have any test. I would strongly discourage that change and encourage people to be coherent in their emails and address book. Emails should not have uppercase.

mlaitinen commented:
> Note: it could have strange result [jacques@exammple.com](mailto:jacques@exammple.com) [jean-jacques@example.com](mailto:jean-jacques@example.com) will both match
> 
> ```python
> In [1]: env['res.partner'].search([('email', '=', 'jacques@example.com')])
> Out[1]: res.partner(10,)
> 
> In [2]: env['res.partner'].search([('email', '=', 'Jacques@example.com')])
> Out[2]: res.partner()
> 
> In [3]: env['res.partner'].search([('email', '=', 'jean-Jacques@example.com')])
> Out[3]: res.partner(11,)
> 
> In [4]: env['res.partner'].search([('email', 'ilike', 'jean-Jacques@example.com')])
> Out[4]: res.partner(11,)
> 
> In [5]: env['res.partner'].search([('email', 'ilike', 'jacques@example.com')])
> Out[5]: res.partner(10, 11)
> ```
> 
> Depending of their order in the database, one or the other will match. It could have catastrophic effects where random users are added to events, synchronization push these changes and nobody understand what is going on. 

I'm a little embarrased now. This has got to be my worst pull request ever. Well spotted. I refactored the solution.

> Moreover your change does not have any test.

Good to see that Odoo these days pays more attention not only to documentation, but to tests as well. I added a test, which also includes testing the capitalization of the attendee's email address. That was already handled properly by this module, but doesn't hurt to include it in the test for consistency reasons.

> I would strongly discourage that change and encourage people to be coherent in their emails and address book. Emails should not have uppercase.

I completely agree that email addressess should always be in lowercase, but sometimes these things just aren't in your hands. I'm glad that whoever wrote the attendee handling code had a friendlier attitude towards the users.

arj-odoo commented on addons/microsoft_calendar/tests/test_create_events.py:
```diff
@@ -111,6 +111,32 @@ def test_create_simple_event_from_outlook_attendee_calendar_and_organizer_does_n
+
+        # act
+        self.organizer_user.with_user(self.organizer_user).sudo()._sync_microsoft_calendar()
+
+        # assert
+        records = self.env["calendar.event"].search([])
+        new_records = (records - existing_records)
+        self.assertEqual(len(new_records), 1)
+        self.assert_odoo_event(new_records, self.expected_odoo_event_from_outlook)
+        self.assertEqual(new_records.user_id, self.organizer_user)
+
```
Shouldn't we use the microsoft_id field to retrieve the new event coming from microsoft instead of searching all events twice ? Normally these tests are run against database with a low amount of records but sometimes it can be useful to run them against copy of production database. In that case, searching for all events will be a real issue.

mlaitinen replied:
I figured that as there are already 14 occurrences of full searches in other tests in the same class, performance hasn't really been the focus here. I can optimize this test if you feel it's necessary, but there will still be 10 full table searches for `calendar.event` and 4 full table searches for `calendar.recurrence`.

arj-odoo replied:
Indeed, you are right. We should create a task to clean that internally at some point.

arj-odoo reviewed (changes requested):
Thank you for the change, the strategy is far better !

I am happy with the change. I only have a suggestion for the test. I would like to avoid searching for all events twice in the test. On large database, the performances would be catastrophic and we don't really need to do it that way. I would use a dedicated record with dedicated microsoft_id to find the new record easily. :-)
